### ISSUE TITLE
Dynamic stake calculation		

### ISSUE THREAD

**Issue by simhe** [mementum/backtrader#101]
Hi.
At the moment I am running a VBA script to calculate an indicator after the market close. With the first open price next morning (the broker offers) I am calculating the number of shares because I want to be invested 100%. The order is opened successfully.

Now I want to test the strategy in backtrader. And the question is how can I calculate the stakes dynamically?

My ugly 'solution':

```
def __init__(self):
         self.dataclose = self.datas[0].close

[...]

def next(self):
   curopen = self.dataclose[0]
   try:
      # next open price is the close price next day
      nextopen = self.dataclose[1]
   except Exception as e:
      nextopen = self.dataclose[0]
   # normally I would except: cash / next open price = stake
   # but backtrader throws margin exception if the current open > next open price
   if curopen < nextopen:
      stake = int(self.broker.getcash() / nextopen)
   else:
      stake = int(self.broker.getcash() / curopen)
```

> Note: I have only close prices, so open = close price

Do you have a better idea?

Maybe in the future you could implement a method `order_target_percent` like in `zipline`? Would be great!

`order_target_percent(symbol('AAPL'), 1.0)`

**Comment by mementum** [comment]
The name `nextopen` and `self.dataclose[1]` show that this idea is not compatible with *backtrader* approach.

`next` is called to evaluate the delivered bar and this is at index `0`. The previous bar is at index `-1` and the one before that one is at index `-2` and so until the initial bar.

Even if for performance reasons the prices under backtesting have been preloaded, looking at index `1` is out of the question because it defeats the purpose: *making an evaluation based on something which is yet to happen*, in this case the next *opening* price.

When `next` is called is called to perform the *next* evaluation (in contrast to the last which was already done). The current and last available price is, again, at index `0`.

Without knowing what exactly `order_target_percent` does behind the scenes, it cannot for sure make a calculation based on a not yet available price (in this case `nextopen`).

The next `open` price can only be known when it happens.

If you use daily bars:

  - `next` evaluates a closed bar and an order set during the evaluation will be executed with the next bar. You can make an educated guess as to how much the *opening* price will gap if you set a market order, but the order can be rejected if too many shares are requested or fall short of coming close to your 100% target.

If you use sub-day bars:

  - `next` will see the open price and you can decide if you enter in the next sub-day bar with a `Market` order (which again may not meet your expectations) or you can set a `Limit` order for an exact 100% investment but fail to be matched at the requested price.

But magic doesn't exist and when the market is closed the `nextopen` price cannot be known.

**Comment by simhe** [comment]
Thanks for your answer. I know it is not compatible but it works and I need this flexibility of a dynamic stake in my portfolio. The results are comparable with the results of my excel sheet and the real trading results. I dont need magic, I only want to test my real trading. The broker offers me after the market is opened a fix price for a few seconds.. with this price I can calculate the number of shares. Thats it. 

I dont think that `order_target_percent` is working like my approach behind the scences but it exists and I think this approach is very common. A lot of strategies have the idea to buy x% of share y, instead of buying a fix amount of shares.

Browsing your closed issues I found issue #72 "Variable Sizer".  Its similiar, but his workaround is not 100% clear for me. In my opinion there is also a margin error possible.

**Comment by mementum** [comment]
> "The broker offers me after the market is opened a fix price for a few seconds.. " 

May the problem be that you test with daily bars but want execution based on intraday events? It seems so.

> "with this price I can calculate the number of shares"

You calculate the number of shares and not the real broker. The same is done in the platform which tries to remain as realistic as possible.

If you have "a fixed price for a few seconds" this is an intraday bar at the opening of the session. During this bar you can calculate the number of shares you can acquire. Or a sizer like in #72 can do it for you, but based on an actual price.

In remaining realistic backtrader cannot look forward for preloaded prices to make a calculation for you, because "preloading" is an artifact intended for performance and not to cheat looking into future prices.

Version 1.5.0.x has added live data feed. Should live data feeds also support giving you a non existent price (the next opening price) to support your strategy?

This seems like an intraday vs daily backtesting issue.

`order_target_percent` is what many banks offer: *tell me how much you want to invest and based on the "ACTUAL" price I calculate the number shares for you". But it needs an "actual" price and not a "future" price.

**Comment by simhe** [comment]
For `backtrader` the actual price is the **close price** one day before (this price is used normally to calculate the number of shares, same as in #72). But the order is executed with the next price, in this case next day in the morning with the **open price** (which is NOW the actual price and this price I need to calculate the number of shares!). So I dont need future prices.. I need the real actual price. 

Let me explain my real trading (long-term strategy, no intraday):
Monday evening:
The broker give me the last price of a share (close price). With this price and all the other close prices the days before I calculate a signal. For example it is telling me to buy. Market is closed, I have to wait.

Tuesday morning:
The broker give me the first price (open price). With this price I can calculate the number of shares and I can open a order without getting a margin error. (**backtrader uses the close price of monday not the actual price and maybe my order failed to open**).

So all data I needed is a close and a open price. No intraday data is needed! In intraday mode I wouldn't get this trouble, thats true. 

What can I do?

1.) Using `self.dataclose[1]` as described. Yes for backtrader it is a future price, but in fact it is the actual price.
2.) Changing to intraday mode, but I dont have intraday data. So I have to use the same price (the close price) for every minute.
3.)...any idea?

**Comment by mementum** [comment]
Sorry, but you have the concepts wrong. When you see a *day* bar is because the bar is **complete** and the day is **over**.

  - Monday: you get *daily* bar (market is closed) and you calculate something based on the "daily" already closed bar. You get a signal. If you issue an order at this point in time, the *opening* price of the next day is unknown, because the day has not yet started.

  - Tuesday: when you see the *daily* bar, just like the bar on Monday, the bar is closed, the day is over. You can no longer buy at the *open* price (unless it happens to repeat at a later point in time)

Your broker gives you the *opening* price in real-time, i.e. intraday. If the broker is so nice as to make a promise with regards to this price and guarantee execution for you it must be a very nice broker. The brokers in this part of the world make no execution guarantee on a price which has already happened. The *opening* price may be the highest price of the day and no longer seen in the entire session (or the lowest)

If you want to do something before Tuesday closes, and especially near to the opening moment you have to do it with intraday prices, because you operate on a non-yet closed daily bar. The daily bar does not exist yet because the day is not yet over.

1) Feel free to use `self.data.close[1]`. You can also use `self.data.close[1000]`. It is exactly so pointless as `1`.

2) Intraday is of course supported, there are several examples of it and it includes *resampling* and *replaying* of intraday data.

3) Use the proper timeframe (intraday) and not the wrong timeframe *day* for the order execution.

**Comment by simhe** [comment]
Thank you. I will try it with intraday.

**Comment by mementum** [comment]
A day to "day in steps" filter which alleviates the need for intraday data by breaking a "daily" bar in 2 bars (an opening bar and the rest), which allows for introduction of orders upon seeing the opening price. See this:

  - http://www.backtrader.com/posts/2016-07-13-day-in-steps/day-in-steps/

**Comment by mementum** [comment]
The sizer interface which would allow for real dynamic stake calculation is available with 1.6.4.93
